compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/cyclefind/GlobalVariableRefAnalyzer.java

```diff
@@ -0,0 +1,356 @@
+        this.pkgNode = pkgNode;
+        this.dlog = dlog;
+        this.globalNodeDependsOn = globalNodeDependsOn;
+        this.dependencyNodes = new HashMap<>();
+        this.cycles = new ArrayList<>();
+        this.nodeInfoStack = new ArrayDeque<>();
+        this.dependencyOrder = new ArrayList<>();
+    }
+
+    /**
+     * Analyze the global variable references and reorder them or emit error if they contain cyclic references.
+     */
+    public void analyzeAndReOrder() {
+        List<BSymbol> globalVarsAndDependentFuncs = getGlobalVariablesAndDependentFunctions();
+
+        pruneDependencyRelations(this.globalNodeDependsOn);
+
+        Set<BSymbol> sorted = new LinkedHashSet<>();
+        LinkedList<BSymbol> dependencies = new LinkedList<>();
+
+        for (BSymbol symbol : globalVarsAndDependentFuncs) {
+            // Only analyze unvisited nodes.
+            // Do DFS into dependency providers to detect cycles.
+            if (!dependencyNodes.containsKey(symbol)) {
+                NodeInfo node = new NodeInfo(curNodeId++, symbol);
+                dependencyNodes.put(symbol, node);
+                analyzeProvidersRecursively(node);
+            }
+            // Extract all the dependencies found in last call to analyzeProvidersRecursively
+            if (!dependencyOrder.isEmpty()) {
+                List<BSymbol> symbolsProvidersOrdered = dependencyOrder.stream()
+                        .map(n -> n.symbol)
+                        .collect(Collectors.toList());
+                dependencies.addAll(symbolsProvidersOrdered);
+                dependencyOrder.clear();
+            }
+
+            List<BSymbol> symbolsProviders = globalNodeDependsOn.get(symbol);
+            boolean symbolHasProviders = symbolsProviders != null && !symbolsProviders.isEmpty();
+
+            // Independent variable declaration, add to sorted list.
+            if (!symbolHasProviders && (!sorted.contains(symbol))) {
+                moveAndAppendToSortedList(symbol, dependencies, sorted);
+            }
+            // Dependent variable, and all the dependencies are satisfied, add to sorted list.
+            if (symbolHasProviders && sorted.containsAll(symbolsProviders) && !sorted.contains(symbol)) {
+                moveAndAppendToSortedList(symbol, dependencies, sorted);
+            }
+
+            // If we can satisfy the dependencies' dependencies then we can add those dependencies to sorted list now.
+            addDependenciesDependencies(dependencies, sorted);
+        }
+        sorted.addAll(dependencies);
+
+        if (cycles.stream().anyMatch(c -> c.size() > 1)) {
+            // Cyclic error found no need to sort.
+            return;
+        }
+
+        projectSortToGlobalVarsList(sorted);
+        projectSortToTopLevelNodesList();
+
+    }
+
+    private void pruneDependencyRelations(Map<BSymbol, List<BSymbol>> globalNodeDependsOn) {
+        List<BSymbol> dependents = new ArrayList<>(globalNodeDependsOn.keySet());
```

Comment:
Cant we iterate `globalNodeDependsOn.keySet()` directly?

Reply:
This keySet coping is to avoid ConcurrentModificationException, since we are modifying the map in this loop.